File: 3.cpp
// Given a string, find the length of the longest substring without repeating
// characters.
//
// Example 1:
//
// Input: "abcabcbb"
// Output: 3
// Explanation: The answer is "abc", with the length of 3.
// Example 2:
//
// Input: "bbbbb"
// Output: 1
// Explanation: The answer is "b", with the length of 1.
// Example 3:
//
// Input: "pwwkew"
// Output: 3
// Explanation: The answer is "wke", with the length of 3.
//              Note that the answer must be a substring, "pwke" is a
//              subsequence and not a substring.

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<int, int> m;
        int start = 0;
        int longest = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (m.count(s[i])) {
                start = max(start, m[s[i]] + 1);
            }
            m[s[i]] = i;
            longest = max(longest, i - start + 1);
        }

        return longest;
    }
};

class Solution2 {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> m(128, 0);
        int start = 0, end = 0;
        int longest = 0;

        while (end < s.size()) {
            m[s[end]]++;
            while (m[s[end]] > 1) {
                if (end - start > longest) {
                    longest = end - start;
                }
                m[s[start]]--;
                start++;
            }
            end++;
        }

        return max(longest, end - start);
    }
};

class Solution3 {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> pos(128, -1);
        int start = -1, end = 0;
        int longest = 0;

        for (int i = 0; i < s.size(); ++i) {
            // we find one character appear again which already within the gap
            // from start to i, update start
            if (pos[s[i]] > start) {
                start = pos[s[i]];
            }
            pos[s[i]] = i;
            longest = max(longest, i - start);
        }

        return longest;
    }
};

int main() {
    string testcase = "abcabcbb";
    string testcase1 = "dvdf";
    Solution s;
    cout << s.lengthOfLongestSubstring(testcase) << endl;

    return 0;
}
/*
 *        d     v     d     f
 *  -1    0     1     2     3
 *
 *  start i
 *
 *
 *        d     v     d     f
 *  -1    0     1     2     3
 *
 *  start       i
 *
 *
 *        d     v     d     f
 *  -1    0     1     2     3
 *
 *  start             i
 *(when d appear again, start should be in previous d position.)
 *
 *
 *        d     v     d     f
 *  -1    0     1     2     3
 *
 *      start         i
 *
 *
 *        d     v     d     f
 *  -1    0     1     2     3
 *
 *       start              i
 */

File: Python/3.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = -1
        pos = {}
        maxLen = 0
        for i in range(len(s)):
            if (s[i] not in pos) or (pos[s[i]] <= start):
                maxLen = max(maxLen, i-start)
            else:
                start = pos[s[i]]
            pos[s[i]] = i
        return maxLen
                
'''
a b c a b c b b
0 1 2 3 4 5 6 7

i=0
start = -1
maxLen = 1
pos{'a':0, }

i=1
start = -1
maxLen = 2
pos{'a':0, 'b':1}

i=2
start = -1
maxLen = 3
pos{'a':0, 'b':1, 'c':2}

i = 3
start = 2
pos{'a':3. 'b':1, 'c':2}


"t m m z u x t"
 0 1 2 3 4 5 6

start = -1
i=0
maxLen = 1
pos{'t':0}

i=1
maxLen = 2
pos{'t':0, 'm':1}

i=2
start = 1
pos{'t':0, 'm':2}

i=3
maxLen = 2
pos{'t':0, 'm':2, 'z':3}

i=4
maxLen = 3
pos{'t':0, 'm':2, 'z':3, 'u':4}

i=5
maxLen = 4
pos{'t':0, 'm':2, 'z':3, 'u':4, 'x':5}

i=6
maxLen = 5
pos{'t':6, 'm':2, 'z':3, 'u':4, 'x':5}

'''
